src/platform/pcp.c

```diff
@@ -0,0 +1,574 @@
+    if (PcpContext == NULL) {
+        QuicTraceEvent(
+            AllocFailure,
+            "Allocation of '%s' failed. (%llu bytes)",
+            "CXPLAT_PCP",
+            PcpContextSize);
+        Status = QUIC_STATUS_OUT_OF_MEMORY;
+        goto Exit;
+    }
+
+    CxPlatZeroMemory(PcpContext, PcpContextSize);
+    PcpContext->ClientContext = Context;
+    PcpContext->ClientCallback = Handler;
+    PcpContext->GatewayCount = GatewayAddressesCount;
+
+    for (uint32_t i = 0; i < GatewayAddressesCount; ++i) {
+        QuicAddrSetPort(&GatewayAddresses[i], CXPLAT_PCP_PORT);
+        Status =
+            CxPlatSocketCreateUdp(
+                Datapath,
+                NULL,
+                &GatewayAddresses[i],
+                PcpContext,
+                CXPLAT_SOCKET_FLAG_PCP,
+                &PcpContext->GatewaySockets[i]);
+        if (QUIC_FAILED(Status)) {
+            goto Exit;
+        }
+    }
+
+    *NewPcpContext = PcpContext;
+    Status = QUIC_STATUS_SUCCESS;
+
+Exit:
+
+    if (QUIC_FAILED(Status)) {
+        if (PcpContext != NULL) {
+            CxPlatPcpUninitialize(PcpContext);
+        }
+    }
+
+    if (GatewayAddresses != NULL) {
+        CXPLAT_FREE(GatewayAddresses, QUIC_POOL_DATAPATH_ADDRESSES);
+    }
+
+    return Status;
+}
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+void
+CxPlatPcpUninitialize(
+    _In_ CXPLAT_PCP* PcpContext
+    )
+{
+    CXPLAT_DBG_ASSERT(PcpContext != NULL);
+
+    for (uint32_t i = 0; i < PcpContext->GatewayCount; ++i) {
+        if (PcpContext->GatewaySockets[i] != NULL) {
+            CxPlatSocketDelete(PcpContext->GatewaySockets[i]);
+        }
+    }
+
+    CXPLAT_FREE(PcpContext, QUIC_POOL_PCP);
+}
+
+_IRQL_requires_max_(DISPATCH_LEVEL)
+void
+CxPlatPcpProcessDatagram(
+    _In_ CXPLAT_PCP* PcpContext,
+    _In_ CXPLAT_RECV_DATA* Datagram
+    )
+{
+    PCP_RESPONSE* Response = (PCP_RESPONSE*)Datagram->Buffer;
+
+    if (Datagram->BufferLength < PCP_MAP_RESPONSE_SIZE) {
+        QuicTraceEvent(
+            LibraryError,
+            "[ lib] ERROR, %s.",
+            "PCP: Invalid length");
+        return;
+    }
+
+    if (Response->Version != PCP_VERSION) {
+        QuicTraceEvent(
+            LibraryError,
+            "[ lib] ERROR, %s.",
+            "PCP: Invalid version");
+        return;
+    }
+
+    if (Response->Request != 1) {
+        QuicTraceEvent(
+            LibraryError,
+            "[ lib] ERROR, %s.",
+            "PCP: Unexpected request");
+        return;
+    }
+
+    CXPLAT_PCP_EVENT Event;
+    CxPlatCopyMemory(Event.FAILURE.Nonce, Response->MAP.MappingNonce, CXPLAT_PCP_NONCE_LENGTH);
+    QUIC_ADDR InternalAddress;
+    CxPlatCopyMemory(&InternalAddress, &Datagram->Tuple->LocalAddress, sizeof(QUIC_ADDR));
+    InternalAddress.Ipv6.sin6_port = Response->MAP.InternalPort;
+    QUIC_ADDR ExternalAddress = {0};
+    QUIC_ADDR RemotePeerAddress = {0};
+
+    if (Response->ResultCode != PCP_RESULT_SUCCESS) {
+        Event.Type = CXPLAT_PCP_EVENT_FAILURE;
```

Comment:
Either we should have an extra space at the top of this if to match the other cases, or remove the space at the top of the other spaces. Your choice.